bellman/src/constraint.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{Namespace, domain, error};
+use domain::{Variable, LinearCombination, Index};
+use error::Result;
+
+use ff::ScalarEngine;
+
+/// Represents a constraint system which can have new variables
+/// allocated and constrains between them formed.
+pub trait ConstraintSystem<E>: Sized 
+where
+    E: ScalarEngine
+{
+    /// Represents the type of the "root" of this constraint system
+    /// so that nested namespaces can minimize indirection.
+    type Root: ConstraintSystem<E>;
+
+    /// Return the "one" input variable
+    fn one() -> Variable {
+        Variable::new_unchecked(Index::Input(0))
+    }
+
+    /// Allocate a private variable in the constraint system. The provided function is used to
+    /// determine the assignment of the variable. The given `annotation` function is invoked
+    /// in testing contexts in order to derive a unique name for this variable in the current
+    /// namespace.
+    fn alloc<F, A, AR>(&mut self, annotation: A, f: F) -> Result<Variable>
+    where
+        F: FnOnce() -> Result<E::Fr>,
+        A: FnOnce() -> AR,
+        AR: Into<String>;
+
+    /// Allocate a public variable in the constraint system. The provided function is used to
+    /// determine the assignment of the variable.
+    fn alloc_input<F, A, AR>(&mut self, annotation: A, f: F) -> Result<Variable>
+    where
+        F: FnOnce() -> Result<E::Fr>,
+        A: FnOnce() -> AR,
+        AR: Into<String>;
+
+    /// Enforce that `A` * `B` = `C`. The `annotation` function is invoked in testing contexts
+    /// in order to derive a unique name for the constraint in the current namespace.
+    fn enforce<A, AR, LA, LB, LC>(&mut self, annotation: A, a: LA, b: LB, c: LC)
+    where
+        A: FnOnce() -> AR,
+        AR: Into<String>,
+        LA: FnOnce(LinearCombination<E>) -> LinearCombination<E>,
+        LB: FnOnce(LinearCombination<E>) -> LinearCombination<E>,
+        LC: FnOnce(LinearCombination<E>) -> LinearCombination<E>;
+
+    /// Create a new (sub)namespace and enter into it. Not intended
+    /// for downstream use; use `namespace` instead.
+    fn push_namespace<NR, N>(&mut self, name_fn: N)
+    where
+        NR: Into<String>,
+        N: FnOnce() -> NR;
+
+    /// Exit out of the existing namespace. Not intended for
+    /// downstream use; use `namespace` instead.
```

Comment:
What is downstream in this context? Unclear

Reply:
Original comment not ours. I think they’re referring to Dow stream as everyone using the library. 

In general, the public interface is quite weak so we will come back to this when we’re reworking the API together